packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,520 @@
+          0.0,
+        ),
+        layoutExtent: max(layoutExtent - constraints.scrollOffset, 0.0),
+      );
+    } else {
+      // If we never started overscrolling, return no geometry.
+      geometry = SliverGeometry.zero;
+    }
+  }
+
+  @override
+  void paint(PaintingContext paintContext, Offset offset) {
+    if (constraints.overlap < 0.0 ||
+        constraints.scrollOffset + child.size.height > 0) {
+      paintContext.paintChild(child, offset);
+    }
+  }
+
+  // Nothing special done here because this sliver always paints its child
+  // exactly between paintOrigin and paintExtent.
+  @override
+  void applyPaintTransform(RenderObject child, Matrix4 transform) {}
+}
+
+/// The current state of the refresh control.
+///
+/// Passed into the [RefreshControlIndicatorBuilder] builder function so
+/// users can show different UI in different modes.
+enum RefreshIndicatorMode {
+  /// Initial state, when not being overscrolled into. Or after the overscroll
+  /// is canceled or after done and the sliver retracted away.
+  inactive,
+  /// While being overscrolled but not far enough yet to trigger the refresh.
+  drag,
+  /// Dragged far enough that the onRefresh callback will run and the dragged
+  /// displacement is not yet at the final refresh resting state.
+  armed,
+  /// While the onRefresh task is running.
+  refresh,
+  /// While the indicator is animating away after refreshing.
+  done,
+}
+
+/// A builder function that can create a different widget to show in the refresh
+/// indicator space depending on the current state of the refresh control and
+/// the space available.
+///
+/// The `refreshTriggerPullDistance`, `refreshIndicatorExtent` parameters are
+/// the same values passed into the [CupertinoRefreshControl].
+///
+/// The `pulledExtent` parameter is the currently available space either from
+/// overscrolling or as held by the sliver during refresh.
+typedef Widget RefreshControlIndicatorBuilder(
+  BuildContext context,
+  RefreshIndicatorMode refreshState,
+  double pulledExtent,
+  double refreshTriggerPullDistance,
+  double refreshIndicatorExtent,
+);
+
+/// A callback function that's invoked when the [CupertinoRefreshControl] is
+/// pulled a `refreshTriggerPullDistance`. Must return a [Future]. Upon
+/// completion of the [Future], the [CupertinoRefreshControl] enters the
+/// [RefreshIndicatorMode.done] state and will start to go away.
+typedef Future<void> RefreshCallback();
+
+/// A sliver widget implementing the iOS-style pull to refresh content control.
+///
+/// When inserted as the first sliver in a scroll view or behind other slivers
+/// that still lets the scrollable overscroll in front of this sliver (such as
+/// the [CupertinoSliverNavigationBar], this widget will:
+///
+///  * Let the user draw inside the overscrolled area via the passed in [builder].
+///  * Trigger the provide [onRefresh] function when overscroll far enough to pass
```

Comment:
provide --> provided
overscroll --> overscrolled

Reply:
that sentence made no sense :S thanks